File: src/main/java/de/pflugradts/pwman3/application/commandhandling/handler/CustomSetCommandHandler.java
package de.pflugradts.pwman3.application.commandhandling.handler;

import com.google.common.eventbus.Subscribe;
import com.google.inject.Inject;
import de.pflugradts.pwman3.application.UserInterfaceAdapterPort;
import de.pflugradts.pwman3.application.commandhandling.command.CustomSetCommand;
import de.pflugradts.pwman3.application.configuration.ReadableConfiguration;
import de.pflugradts.pwman3.application.failurehandling.FailureCollector;
import de.pflugradts.pwman3.domain.model.transfer.Bytes;
import de.pflugradts.pwman3.domain.model.transfer.Input;
import de.pflugradts.pwman3.domain.model.transfer.Output;
import de.pflugradts.pwman3.domain.service.password.PasswordService;

public class CustomSetCommandHandler implements CommandHandler {

    @Inject
    private ReadableConfiguration configuration;
    @Inject
    private FailureCollector failureCollector;
    @Inject
    private PasswordService passwordService;
    @Inject
    private UserInterfaceAdapterPort userInterfaceAdapterPort;

    @Subscribe
    private void handleCustomSetCommand(final CustomSetCommand customSetCommand) {
        if (commandConfirmed(customSetCommand)) {
            final var secureInput = userInterfaceAdapterPort
                    .receiveSecurely(Output.of(Bytes.of("Enter custom password: ")))
                    .onFailure(failureCollector::collectInputFailure)
                    .getOrElse(Input.empty());
            if (secureInput.isEmpty()) {
                userInterfaceAdapterPort.send(Output.of(Bytes.of("Empty input - Operation aborted.")));
            } else {
                passwordService.putPasswordEntry(customSetCommand.getArgument(), secureInput.getBytes());
            }
            secureInput.invalidate();
        } else {
            userInterfaceAdapterPort.send(Output.of(Bytes.of("Operation aborted.")));
        }
        customSetCommand.invalidateInput();
        userInterfaceAdapterPort.sendLineBreak();
    }

    private boolean commandConfirmed(final CustomSetCommand customSetCommand) {
        if (configuration.getApplication().getPassword().isPromptOnRemoval()
                && passwordService.entryExists(customSetCommand.getArgument())
                .onFailure(throwable -> failureCollector
                        .collectPasswordEntryFailure(customSetCommand.getArgument(), throwable))
                .getOrElse(false)) {
            return userInterfaceAdapterPort
                    .receiveConfirmation(Output.of(Bytes.of(String.format(
                            "Existing Password Entry '%s' will be irrevocably overwritten.%n"
                                    + "Input 'c' to confirm or anything else to abort.%nYour input: ",
                            customSetCommand.getArgument().asString()))));
        }
        return true;
    }

}

File: src/main/kotlin/de/pflugradts/passbird/application/commandhandling/handler/egg/CustomSetCommandHandler.kt
package de.pflugradts.passbird.application.commandhandling.handler.egg
import com.google.common.eventbus.Subscribe
import de.pflugradts.passbird.application.SecureInputUnavailableException
import de.pflugradts.passbird.application.UserInterfaceAdapterPort
import de.pflugradts.passbird.application.commandhandling.CommandExecutionTracker
import de.pflugradts.passbird.application.commandhandling.command.CustomSetCommand
import de.pflugradts.passbird.application.commandhandling.handler.CommandHandler
import de.pflugradts.passbird.application.configuration.ReadableConfiguration
import de.pflugradts.passbird.domain.model.egg.InvalidEggIdException
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.shellOf
import de.pflugradts.passbird.domain.model.transfer.Input
import de.pflugradts.passbird.domain.model.transfer.Output.Companion.outputOf
import de.pflugradts.passbird.domain.model.transfer.OutputFormatting.OPERATION_ABORTED
import de.pflugradts.passbird.domain.service.password.PasswordService
import de.pflugradts.passbird.domain.service.password.PasswordService.EggNotExistsAction.DO_NOTHING
class CustomSetCommandHandler constructor(
    private val configuration: ReadableConfiguration,
    private val passwordService: PasswordService,
    private val userInterfaceAdapterPort: UserInterfaceAdapterPort,
    private val commandExecutionTracker: CommandExecutionTracker,
) : CommandHandler {
    @Subscribe
    private fun handleCustomSetCommand(customSetCommand: CustomSetCommand) {
        if (commandConfirmed(customSetCommand)) {
            processConfirmedCustomSetCommand(customSetCommand)
        } else {
            commandExecutionTracker.markAborted()
            userInterfaceAdapterPort.send(outputOf(shellOf("Operation aborted."), OPERATION_ABORTED))
        }
        customSetCommand.invalidateInput()
        userInterfaceAdapterPort.sendLineBreak()
    }
    private fun processConfirmedCustomSetCommand(customSetCommand: CustomSetCommand) {
        try {
            passwordService.challengeEggId(customSetCommand.argument)
            receiveCustomPassword()?.let { secureInput ->
                try {
                    if (secureInput.isEmpty) {
                        commandExecutionTracker.markAborted()
                        userInterfaceAdapterPort.send(outputOf(shellOf("Empty input - Operation aborted."), OPERATION_ABORTED))
                    } else if (passwordService.putEgg(customSetCommand.argument, secureInput.shell).failure) {
                        commandExecutionTracker.markFailure()
                    }
                } finally {
                    secureInput.invalidate()
                }
            }
        } catch (ex: InvalidEggIdException) {
            commandExecutionTracker.markAborted()
            userInterfaceAdapterPort.send(outputOf(shellOf("${ex.message} - Operation aborted."), OPERATION_ABORTED))
        }
    }
    private fun receiveCustomPassword(): Input? = try {
        userInterfaceAdapterPort.receiveSecurely(outputOf(shellOf("Enter custom Password: ")))
    } catch (_: SecureInputUnavailableException) {
        commandExecutionTracker.markAborted()
        userInterfaceAdapterPort.send(outputOf(shellOf("Operation aborted."), OPERATION_ABORTED))
        null
    }
    private fun commandConfirmed(customSetCommand: CustomSetCommand) =
        if (configuration.application.password.promptOnRemoval && passwordService.eggExists(customSetCommand.argument, DO_NOTHING)) {
            userInterfaceAdapterPort
                .receiveConfirmation(
                    outputOf(
                        shellOf(
                            "Existing Egg '${customSetCommand.argument.asString()}' will be irrevocably overwritten.\n" +
                                "Input 'c' to confirm or anything else to abort.\nYour input: ",
                        ),
                    ),
                )
        } else {
            true
        }
}
